### PR TITLE
Higlass View Config Page Inline Editing Improvements

### PR DESCRIPTION
**Pull request by utku-ozturk**
Pls handle this PR w/ https://github.com/4dn-dcic/shared-portal-components/pull/34

PR is about two main topics: HiGlass Files Table will have a new Experiment Sets column and the HiGlass View Config View Page will have inline editing capabilities:

### A. New Experiment Sets Column: (from Trello)
1. A calculated property called `source` array of linkTo:Item which collapses the revlinked experiment or experimentset. (no embedded fields, just a link.) Check with wranglers to confirm they like the idea. (Alternatively, source could always point to expsets, tracing the revlink from experiment.)
2. add the new source source column to the file search result table.

**solution:** calculated property `source_experiment_sets` list (instead of `source`) added to file type and a column added to Higlass files table.

Screen Shot 2020-07-02 at 15 48 22

### B. Inline Editing (from Trello)
1. Users can edit title, description, and display height if they own the display.
2. But, the edit space is hard to find, going to the edit space wipes unsaved changes to viewconf, and users who can't edit the display can't edit the display height live (even though they can edit the viewconf live.)
3. It would be useful to have boxes to be able to edit these 3 fields that live in the state. Then with clone or save, they would be written to db. Same experience as with viewconf edits.

**solution:** 
1. **Page Title:** inline editing added, data is persistent immediately
2. **Page Description:** inline editing added, data is persistent immediately
3. **Track Title:** inline editing, data is persistent after Save or Clone buttons clicked
4. **Track Width/Height:** data is persistent after Save or Clone buttons clicked
5. **Display height:** inline editing added, data is persistent immediately

**Screenshot - Inline Editing (Title):**
Screen Shot 2020-07-02 at 15 56 13

**Screenhost - Track Editing Is Not Persistent Yet Warning:**
Screen Shot 2020-07-02 at 16 04 16

### CODE REVIEW

**Comment by j1z0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 8
- Added 5
           

Complexity increasing per file
==============================
- src/encoded/static/components/PageTitle.js  3
         

Clones added
============
- src/encoded/static/components/item-pages/HiGlassViewConfigView.js  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828&bid=18964932)

File: src/encoded/static/components/item-pages/HiGlassViewConfigView.js [hunk 1]
```diff
@@ -178,34 +186,153 @@ export class HiGlassViewConfigTabView extends React.PureComponent {
...
-        return !!(session && _.findWhere(actions, { 'name' : 'edit' }));
+    setInstanceHeight(instance) {
+        const { context, isFullscreen, windowHeight } = this.props;
+        if (instance === undefined) {
+            let hiGlassComponentHeight;
+            if (isFullscreen) {
+                hiGlassComponentHeight = windowHeight - 120;
+            }
+            else {
+                if (this.state.instance_height > 0) {
```
Comment (j1z0) [inline src/encoded/static/components/item-pages/HiGlassViewConfigView.js:218]: Codacy found an issue: [Must use destructuring state assignment](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/PageTitle.js [hunk 2]
```diff
@@ -154,8 +157,8 @@ const StaticPageTitle = React.memo(function StaticPageTitle(props){
 
 /** Based on 4DN content views & metadata, to be updated re: CGAP */
 const GenericItemPageTitle = React.memo(function GenericItemPageTitle(props){
-    const { context, schemas, alerts, href, session } = props;
-    const itemTitle = object.itemUtil.getTitleStringFromContext(context);
+    const { context, schemas, alerts, href, session, className } = props;
```
Comment (j1z0) [inline src/encoded/static/components/PageTitle.js:160]: Codacy found an issue: ['className' is assigned a value but never used.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js [hunk 3]
```diff
@@ -238,33 +239,42 @@ export class TopRow extends React.Component {
...
 
-    shouldComponentUpdate(nextProps){
+    static defaultProps = {
+        'isInlineEditable': false
+    };
+
+    shouldComponentUpdate(nextProps) {
         if ((nextProps.context) && (!this.props.context || this.props.context.description !== nextProps.context.description)) return true;
         if (nextProps.windowWidth !== this.props.windowWidth) return true;
+        if (nextProps.context.actions !== this.props.context.actions) return true;
```
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:249]: Codacy found an issue: [Must use destructuring props assignment](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js [hunk 4]
```diff
@@ -8,6 +8,7 @@ import queryString from 'query-string';
 import { console, object, schemaTransforms, memoizedUrlParse } from '@hms-dbmi-bgm/shared-portal-components/es/components/util';
 import { LocalizedTime } from '@hms-dbmi-bgm/shared-portal-components/es/components/ui/LocalizedTime';
 import { FlexibleDescriptionBox } from '@hms-dbmi-bgm/shared-portal-components/es/components/ui/FlexibleDescriptionBox';
+import { EditableField, FieldSet } from '@hms-dbmi-bgm/shared-portal-components/es/components/forms/components/EditableField';
```
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:11]: Codacy found an issue: ['EditableField' is defined but never used.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/PageTitle.js [hunk 5]
```diff
@@ -15,6 +15,8 @@ import { content_views } from './globals';
 import { typedefs } from './util';
 import QuickInfoBar from './viz/QuickInfoBar';
 import jsonScriptEscape from './../libs/jsonScriptEscape';
+import { EditableField, FieldSet } from '@hms-dbmi-bgm/shared-portal-components/es/components/forms/components/EditableField';
+import { HiGlassViewConfigTabView } from './../components/item-pages/HiGlassViewConfigView';
```
Comment (j1z0) [inline src/encoded/static/components/PageTitle.js:19]: Codacy found an issue: ['HiGlassViewConfigTabView' is defined but never used.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

**Comment by j1z0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 7
           

Complexity increasing per file
==============================
- src/encoded/static/components/PageTitle.js  3
         

Clones added
============
- src/encoded/static/components/item-pages/HiGlassViewConfigView.js  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828&bid=18964932)

File: src/encoded/static/components/item-pages/HiGlassViewConfigView.js (hunk 1 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/HiGlassViewConfigView.js:218]: Codacy found an issue: [Must use destructuring state assignment](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js (hunk 4 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:11]: Codacy found an issue: ['EditableField' is defined but never used.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/PageTitle.js (hunk 5 above)
Comment (j1z0) [inline src/encoded/static/components/PageTitle.js:19]: Codacy found an issue: ['HiGlassViewConfigTabView' is defined but never used.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js [hunk 6]
```diff
@@ -236,48 +237,65 @@ export class TopRow extends React.Component {
...
-        windowWidth
-    } = props;
+    static defaultProps = {
+        'isInlineEditable': false
+    };
 
-    if (children) {
-        return <div className="item-page-heading">{ children }</div>;
+    shouldComponentUpdate(nextProps) {
+        if ((nextProps.context) && (!this.props.context || this.props.context.description !== nextProps.context.description)) return true;
```
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:247]: Codacy found an issue: [Must use destructuring props assignment (react/destructuring-assignment)](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js [hunk 7]
```diff
@@ -236,48 +237,65 @@ export class TopRow extends React.Component {
...
+        if ((nextProps.context) && (!this.props.context || this.props.context.actions !== nextProps.context.actions)) return true;
```
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:248]: Codacy found an issue: [Must use destructuring props assignment](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js [hunk 8]
```diff
@@ -236,48 +237,65 @@ export class TopRow extends React.Component {
...
+        if (nextProps.windowWidth !== this.props.windowWidth) return true;
```
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:249]: Codacy found an issue: [Must use destructuring props assignment (react/destructuring-assignment)](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/PageTitle.js (hunk 2 above)
Comment (j1z0) [inline src/encoded/static/components/PageTitle.js:160]: Codacy found an issue: ['className' is assigned a value but never used.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

**Comment by j1z0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 7
           

Complexity increasing per file
==============================
- src/encoded/static/components/PageTitle.js  3
         

Clones added
============
- src/encoded/static/components/item-pages/HiGlassViewConfigView.js  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828&bid=18964932)

File: src/encoded/static/components/item-pages/components/ItemHeader.js (hunk 8 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:249]: Codacy found an issue: [Must use destructuring props assignment (react/destructuring-assignment)](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js (hunk 6 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:247]: Codacy found an issue: [Must use destructuring props assignment (react/destructuring-assignment)](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js (hunk 4 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:11]: Codacy found an issue: ['EditableField' is defined but never used.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/HiGlassViewConfigView.js (hunk 1 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/HiGlassViewConfigView.js:218]: Codacy found an issue: [Must use destructuring state assignment](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/PageTitle.js (hunk 2 above)
Comment (j1z0) [inline src/encoded/static/components/PageTitle.js:160]: Codacy found an issue: ['className' is assigned a value but never used.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/HiGlassViewConfigView.js [hunk 9]
```diff
@@ -20,20 +20,19 @@ import { CollapsibleItemViewButtonToolbar } from './components/CollapsibleItemVi
...
         const itemActionsDescriptions = {
-            'edit' : 'Edit the properties of this Item.',
+            'edit': 'Edit the properties of this Item.',
         };
 
         return (
             <ItemHeaderWrapper {..._.pick(this.props, 'context', 'href', 'schemas', 'windowWidth')}>
                 <TopRow typeInfo={this.typeInfo()} itemActionsDescriptions={itemActionsDescriptions} />
-                <MiddleRow />
+                <MiddleRow isInlineEditable={true} />
```
Comment (alexkb0009) [inline src/encoded/static/components/item-pages/HiGlassViewConfigView.js:35]: (Not Important) Can also just have `<MiddleRow isInlineEditable />` which is analogous to `<MiddleRow isInlineEditable={true} />`

**Comment by j1z0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 4
           

Complexity increasing per file
==============================
- src/encoded/static/components/PageTitle.js  3
         

Clones added
============
- src/encoded/static/components/item-pages/HiGlassViewConfigView.js  9
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828&bid=18964932)

File: src/encoded/static/components/PageTitle.js [hunk 10]
```diff
@@ -154,8 +156,8 @@ const StaticPageTitle = React.memo(function StaticPageTitle(props){
 
 /** Based on 4DN content views & metadata, to be updated re: CGAP */
 const GenericItemPageTitle = React.memo(function GenericItemPageTitle(props){
-    const { context, schemas, alerts, href, session } = props;
-    const itemTitle = object.itemUtil.getTitleStringFromContext(context);
+    const { context, schemas, alerts, href, session, className } = props;
```
Comment (j1z0) [inline src/encoded/static/components/PageTitle.js:159]: Codacy found an issue: ['className' is assigned a value but never used.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js [hunk 11]
```diff
@@ -236,48 +236,65 @@ export class TopRow extends React.Component {
...
-        windowWidth
-    } = props;
+    static defaultProps = {
+        'isInlineEditable': false
+    };
 
-    if (children) {
-        return <div className="item-page-heading">{ children }</div>;
+    shouldComponentUpdate(nextProps) {
+        if ((nextProps.context) && (!this.props.context || this.props.context.description !== nextProps.context.description)) return true;
```
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:246]: Codacy found an issue: [Must use destructuring props assignment (react/destructuring-assignment)](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js [hunk 12]
```diff
@@ -236,48 +236,65 @@ export class TopRow extends React.Component {
...
+        if ((nextProps.context) && (!this.props.context || this.props.context.actions !== nextProps.context.actions)) return true;
+        if (nextProps.windowWidth !== this.props.windowWidth) return true;
```
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:248]: Codacy found an issue: [Must use destructuring props assignment (react/destructuring-assignment)](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js [hunk 13]
```diff
@@ -236,48 +236,65 @@ export class TopRow extends React.Component {
...
-    } = props;
+    static defaultProps = {
+        'isInlineEditable': false
+    };
 
-    if (children) {
-        return <div className="item-page-heading">{ children }</div>;
+    shouldComponentUpdate(nextProps) {
+        if ((nextProps.context) && (!this.props.context || this.props.context.description !== nextProps.context.description)) return true;
+        if ((nextProps.context) && (!this.props.context || this.props.context.actions !== nextProps.context.actions)) return true;
```
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:247]: Codacy found an issue: [Must use destructuring props assignment](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

**Comment by utku-ozturk** [comment]
@alexkb0009 thanks for your detailed review.

1. you are right, adding the fadein effect is like reinventing the wheel. I removed both @keyframes in scss and markup in js file.
2. Sure. I'll add some comments shortly.
3. Yes, right. But let's add the updated (or new) version of EditableField after the PR is merged as a new feature, since HiGlassViewConfig inline editing feature has been waiting for a very long time.

**Comment by j1z0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 4
           

Complexity increasing per file
==============================
- src/encoded/static/components/PageTitle.js  3
         

Clones removed
==============
+ src/encoded/static/components/item-pages/HiGlassViewConfigView.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828&bid=18964932)

File: src/encoded/static/components/PageTitle.js (hunk 10 above)
Comment (j1z0) [inline src/encoded/static/components/PageTitle.js:159]: Codacy found an issue: ['className' is assigned a value but never used.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js (hunk 11 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:246]: Codacy found an issue: [Must use destructuring props assignment (react/destructuring-assignment)](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js (hunk 12 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:248]: Codacy found an issue: [Must use destructuring props assignment (react/destructuring-assignment)](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js (hunk 13 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:247]: Codacy found an issue: [Must use destructuring props assignment](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

**Comment by j1z0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 4
           

Complexity increasing per file
==============================
- src/encoded/static/components/PageTitle.js  3
         

Clones removed
==============
+ src/encoded/static/components/item-pages/HiGlassViewConfigView.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828&bid=18964932)

File: src/encoded/static/components/PageTitle.js (hunk 10 above)
Comment (j1z0) [inline src/encoded/static/components/PageTitle.js:159]: Codacy found an issue: ['className' is assigned a value but never used.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js (hunk 11 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:246]: Codacy found an issue: [Must use destructuring props assignment (react/destructuring-assignment)](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js (hunk 12 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:248]: Codacy found an issue: [Must use destructuring props assignment (react/destructuring-assignment)](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js (hunk 13 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:247]: Codacy found an issue: [Must use destructuring props assignment](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

**Comment by j1z0** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 4
           

Complexity increasing per file
==============================
- src/encoded/static/components/PageTitle.js  3
         

Clones removed
==============
+ src/encoded/static/components/item-pages/HiGlassViewConfigView.js  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828&bid=18964932)

File: src/encoded/static/components/PageTitle.js (hunk 10 above)
Comment (j1z0) [inline src/encoded/static/components/PageTitle.js:159]: Codacy found an issue: ['className' is assigned a value but never used.](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js (hunk 11 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:246]: Codacy found an issue: [Must use destructuring props assignment (react/destructuring-assignment)](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js (hunk 12 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:248]: Codacy found an issue: [Must use destructuring props assignment (react/destructuring-assignment)](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

File: src/encoded/static/components/item-pages/components/ItemHeader.js (hunk 13 above)
Comment (j1z0) [inline src/encoded/static/components/item-pages/components/ItemHeader.js:247]: Codacy found an issue: [Must use destructuring props assignment](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5793828)

**Pull request merged**